x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StGeotile.java

```diff
@@ -0,0 +1,362 @@
+    static void fromFieldAndFieldAndLiteral(
+        BytesRefBlock.Builder results,
+        int p,
+        BytesRefBlock in,
+        int precision,
+        @Fixed Rectangle bounds
+    ) {
+        fromWKB(results, p, in, precision, bounds);
+    }
+
+    @Evaluator(extraName = "FromFieldDocValuesAndFieldAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static void fromFieldDocValuesAndFieldAndLiteral(
+        BytesRefBlock.Builder results,
+        int p,
+        LongBlock encoded,
+        int precision,
+        @Fixed Rectangle bounds
+    ) {
+        fromEncodedLong(results, p, encoded, precision, bounds);
+    }
+
+    @Evaluator(extraName = "FromLiteralAndFieldAndLiteral", warnExceptions = { IllegalArgumentException.class })
+    static BytesRef fromLiteralAndFieldAndLiteral(@Fixed BytesRef in, int precision, @Fixed Rectangle bounds) {
+        return calculateGeotile(GEO.wkbAsPoint(in), precision, bounds);
+    }
+
+    private static void fromWKB(BytesRefBlock.Builder results, int position, BytesRefBlock wkbBlock, int precision) {
+        int valueCount = wkbBlock.getValueCount(position);
+        if (valueCount < 1) {
+            results.appendNull();
+        } else {
+            final BytesRef scratch = new BytesRef();
+            final int firstValueIndex = wkbBlock.getFirstValueIndex(position);
+            if (valueCount == 1) {
+                results.appendBytesRef(calculateGeotile(GEO.wkbAsPoint(wkbBlock.getBytesRef(firstValueIndex, scratch)), precision));
+            } else {
+                results.beginPositionEntry();
+                for (int i = 0; i < valueCount; i++) {
+                    results.appendBytesRef(calculateGeotile(GEO.wkbAsPoint(wkbBlock.getBytesRef(firstValueIndex + i, scratch)), precision));
+                }
+                results.endPositionEntry();
+            }
+        }
+    }
+
+    private static void fromEncodedLong(BytesRefBlock.Builder results, int position, LongBlock encoded, int precision) {
+        int valueCount = encoded.getValueCount(position);
+        if (valueCount < 1) {
+            results.appendNull();
+        } else {
+            final int firstValueIndex = encoded.getFirstValueIndex(position);
+            if (valueCount == 1) {
+                results.appendBytesRef(calculateGeotile(GEO.longAsPoint(encoded.getLong(firstValueIndex)), precision));
+            } else {
+                results.beginPositionEntry();
+                for (int i = 0; i < valueCount; i++) {
+                    results.appendBytesRef(calculateGeotile(GEO.longAsPoint(encoded.getLong(firstValueIndex + i)), precision));
+                }
+                results.endPositionEntry();
+            }
+        }
+    }
+
+    protected static BytesRef calculateGeotile(Point point, int precision, Rectangle bounds) {
+        // For points, filtering the point is as good as filtering the tile
```

Comment:
I don't think this statement is correct. See https://github.com/elastic/elasticsearch/blob/368d47131f2d74742ae89005680352389387bdb8/server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileBoundedPredicate.java

Reply:
I agree. The current code was a quick-and dirty placeholder for the draft PR. I'll definitely switch over to the better code you suggest.